caas/kubernetes/provider/k8s.go

```diff
@@ -0,0 +1,184 @@
+var logger = loggo.GetLogger("juju.kubernetes.provider")
+
+// TODO(caas) should be using a juju specific namespace
+const namespace = "default"
+
+// TODO(caas) - add unit tests
+
+type kubernetesClient struct {
+	*kubernetes.Clientset
+}
+
+// NewK8sProvider returns a kubernetes client for the specified cloud.
+func NewK8sProvider(cloudSpec environs.CloudSpec) (caas.Broker, error) {
+	config, err := newK8sConfig(cloudSpec)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	client, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return &kubernetesClient{client}, nil
+}
+
+func newK8sConfig(cloudSpec environs.CloudSpec) (*rest.Config, error) {
+	if cloudSpec.Credential == nil {
+		return nil, errors.Errorf("cloud %v has no credential", cloudSpec.Name)
+	}
+
+	var CAData []byte
+	for _, cacert := range cloudSpec.CACertificates {
+		CAData = append(CAData, cacert...)
+	}
+
+	credentialAttrs := cloudSpec.Credential.Attributes()
+	return &rest.Config{
+		Host:     cloudSpec.Endpoint,
+		Username: credentialAttrs["Username"],
+		Password: credentialAttrs["Password"],
+		TLSClientConfig: rest.TLSClientConfig{
+			CertData: []byte(credentialAttrs["ClientCertificateData"]),
+			KeyData:  []byte(credentialAttrs["ClientKeyData"]),
+			CAData:   CAData,
+		},
+	}, nil
+}
+
+// EnsureOperator creates a new operator for appName if it doesn't exist.
+func (k *kubernetesClient) EnsureOperator(appName, agentPath string, newConfig caas.NewOperatorConfigFunc) error {
+	if exists, err := k.operatorExists(appName); err != nil {
+		return errors.Trace(err)
+	} else if exists {
+		logger.Debugf("%s operator already deployed", appName)
+		return nil
+	}
+	logger.Infof("deploying %s operator", appName)
+
+	configMapName, err := k.ensureConfigMap(appName, newConfig)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return k.deployOperator(appName, agentPath, configMapName)
+}
+
+func (k *kubernetesClient) ensureConfigMap(appName string, newConfig caas.NewOperatorConfigFunc) (string, error) {
+	mapName := podName(appName) + "-config"
+
+	exists, err := k.configMapExists(mapName)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	if exists {
+		logger.Infof("ConfigMap %s already exists", mapName)
+	} else {
+		config, err := newConfig(appName)
+		if err != nil {
+			return "", errors.Annotate(err, "creating config")
+		}
+		if err := k.createConfigMap(mapName, config); err != nil {
+			return "", errors.Annotate(err, "creating ConfigMap")
+		}
+	}
+	return mapName, nil
+}
+
+func (k *kubernetesClient) configMapExists(configMapName string) (bool, error) {
+	_, err := k.CoreV1().ConfigMaps(namespace).Get(configMapName)
+	if k8serrors.IsNotFound(err) {
+		return false, nil
+	} else if err != nil {
+		return false, errors.Trace(err)
+	}
+	return true, nil
+}
+
+func (k *kubernetesClient) createConfigMap(configMapName string, config []byte) error {
+	_, err := k.CoreV1().ConfigMaps(namespace).Create(&v1.ConfigMap{
```

Comment:
TODO(caas) use secrets for storing agent password?

Reply:
done